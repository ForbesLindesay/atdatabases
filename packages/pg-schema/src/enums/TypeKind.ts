enum TypeKind {
  // N.B. Array is not really a type in postgres,
  // instead Array is just a special kind of base
  // type
  Array = 'array',
  Base = 'b',
  // (e.g., a table's row type)
  Composite = 'c',
  // domains alias other types, with optional constraints
  Domain = 'd',
  Enum = 'e',
  Pseudo = 'p',
}

export default TypeKind;
