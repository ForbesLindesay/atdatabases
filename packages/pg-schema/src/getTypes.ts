import {Connection, sql} from '@databases/pg';
import TypeCateogry from './enums/TypeCategory';
import TypeKind from './enums/TypeKind';
import getAttributes, {Attribute} from './getAttributes';
import getEnumValues from './getEnumValues';

export interface TypeQuery {
  schemaID?: number;
  schemaName?: string;
  typeID?: number;
  typeName?: string;
  category?: TypeCateogry;
}
export interface TypeRecord {
  schemaID: number;
  schemaName: string;
  typeID: number;
  typeName: string;
  kind: TypeKind;
  category: TypeCateogry;
  /**
   * classID if composite type
   */
  classID?: number;
  subtypeID?: number;
  subtypeName?: string;
  basetypeID?: number;
  basetypeName?: string;
  comment: string | null;
}

export interface TypeBase {
  schemaID: number;
  schemaName: string;
  typeID: number;
  typeName: string;
  kind: TypeKind;
  category: TypeCateogry;
  comment: string | null;
}

export interface ArrayType extends TypeBase {
  kind: TypeKind.Array;
  subtypeID: number;
  subtypeName: string;
}
export interface BaseType extends TypeBase {
  kind: TypeKind.Base;
  subtypeID?: number;
  subtypeName?: string;
}
export interface CompositeType extends TypeBase {
  kind: TypeKind.Composite;
  classID: number;
  attributes: Attribute[];
}
export interface DomainType extends TypeBase {
  kind: TypeKind.Domain;
  basetypeID: number;
  basetypeName: string;
}
export interface EnumType extends TypeBase {
  kind: TypeKind.Enum;
  values: string[];
}
export interface PseudoType extends TypeBase {
  kind: TypeKind.Pseudo;
}

export type Type =
  | ArrayType
  | BaseType
  | CompositeType
  | DomainType
  | EnumType
  | PseudoType;

export default async function getTypes(
  connection: Connection,
  query: TypeQuery = {},
): Promise<Type[]> {
  const conditions = typeQuery(query);

  const typeRecords = (await connection.query(sql`
    SELECT
      ns.oid AS "schemaID",
      ns.nspname AS "schemaName",
      ty.oid as "typeID",
      ty.typname AS "typeName",
      ty.typtype AS "kind",
      ty.typcategory AS "category",
      ty.typrelid AS "classID",
      subt.oid as "subtypeID",
      subt.typname AS "subtypeName",
      baset.oid as "basetypeID",
      baset.typname AS "basetypeName",
      
      obj_description(ty.oid, 'pg_type') as "comment"
    FROM pg_catalog.pg_type ty
    INNER JOIN pg_catalog.pg_namespace ns
      ON (ty.typnamespace = ns.oid)
    LEFT OUTER JOIN pg_catalog.pg_type subt
      ON (ty.typelem = subt.oid)
    LEFT OUTER JOIN pg_catalog.pg_type baset
      ON (ty.typbasetype = baset.oid)
    ${conditions.length ? sql`WHERE ${sql.join(conditions, ' AND ')}` : sql``}
  `)) as TypeRecord[];

  return Promise.all(
    typeRecords.map(
      async (tr): Promise<Type> => {
        const base: TypeBase = {
          schemaID: tr.schemaID,
          schemaName: tr.schemaName,
          typeID: tr.typeID,
          typeName: tr.typeName,
          kind: tr.kind,
          category: tr.category,
          comment: tr.comment,
        };
        switch (base.kind) {
          case TypeKind.Array:
          case TypeKind.Base:
            if (tr.category === TypeCateogry.Array) {
              return {
                ...base,
                kind: TypeKind.Array,
                subtypeID: tr.subtypeID!,
                subtypeName: tr.subtypeName!,
              };
            } else {
              return {
                ...base,
                kind: TypeKind.Base,
                subtypeID: tr.subtypeID,
                subtypeName: tr.subtypeName,
              };
            }
          case TypeKind.Composite:
            return {
              ...base,
              kind: TypeKind.Composite,
              classID: tr.classID!,
              attributes: await getAttributes(connection, {
                classID: tr.classID!,
              }),
            };
          case TypeKind.Domain:
            return {
              ...base,
              kind: TypeKind.Domain,
              basetypeID: tr.basetypeID!,
              basetypeName: tr.basetypeName!,
            };
          case TypeKind.Enum:
            return {
              ...base,
              kind: TypeKind.Enum,
              values: (await getEnumValues(connection, {
                typeID: tr.typeID,
              })).map(v => v.value),
            };
          case TypeKind.Pseudo:
            return {
              ...base,
              kind: TypeKind.Pseudo,
            };
          default:
            const kind: never = base.kind;
            return {
              ...base,
              kind,
            };
        }
      },
    ),
  );
}

export function typeQuery(query: TypeQuery) {
  const conditions = [];
  if (query.schemaID) {
    conditions.push(sql`ns.oid = ${query.schemaID}`);
  }
  if (query.schemaName) {
    conditions.push(sql`ns.nspname = ${query.schemaName}`);
  }
  if (query.typeID) {
    conditions.push(sql`ty.oid = ${query.typeID}`);
  }
  if (query.typeName) {
    conditions.push(sql`ty.typname = ${query.typeName}`);
  }
  if (query.category) {
    conditions.push(sql`ty.typcategory = ${query.category}`);
  }
  return conditions;
}
