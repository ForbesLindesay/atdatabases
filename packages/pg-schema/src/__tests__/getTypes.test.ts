import connect, {sql, DataTypeID} from '@databases/pg';
import getTypes, {Type} from '../getTypes';
import TypeCateogry from '../enums/TypeCategory';
import {readFileSync, writeFileSync} from 'fs';
import TypeKind from '../enums/TypeKind';

const db = connect();

// int8 default for pg-promise is string
// but @databases/pg uses number

const INTERVAL =
  '{years: number,months: number,days: number,hours: number,minutes: number,seconds: number,milliseconds: number, toPostgres: () => string, toISO: () => string}';

const typeMappings: {[key in DataTypeID]?: string} = {
  [DataTypeID.bool]: 'boolean',
  [DataTypeID.bytea]: 'Buffer',
  [DataTypeID.circle]: '{x: number, y: number, radius: number}',
  [DataTypeID.float4]: 'number',
  [DataTypeID.int2]: 'number',
  [DataTypeID.int8]: 'number',
  [DataTypeID.interval]: INTERVAL,
  [DataTypeID.json]: 'any',
  [DataTypeID.jsonb]: 'any',
  [DataTypeID.point]: '{x: number, y: number}',
  [DataTypeID.timestamp]: 'Date',
  [DataTypeID._bool]: 'Array<boolean | nul>',
  [DataTypeID._bytea]: 'Array<Buffer | null>',
  [DataTypeID._int2]: 'Array<number | null>',
  [DataTypeID._int8]: 'Array<number | null>',
  [DataTypeID._interval]: `Array<${INTERVAL} | null>`,
  [DataTypeID._json]: 'Array<any | null>',
  [DataTypeID._numeric]: 'Array<number | null>',
  [DataTypeID._point]: 'Array<{x: number, y: number} | null>',
  [DataTypeID._text]: 'Array<string | null>',
  [DataTypeID._timestamp]: 'Array<Date | null>',
};

function writeIfDifferent(filename: string, content: string) {
  try {
    if (readFileSync(filename, 'utf8') === content) {
      return;
    }
  } catch (ex) {
    if (ex.code !== 'ENOENT') throw ex;
  }
  writeFileSync(filename, content);
}

test('get built in types', async () => {
  const builtInTypes = await getTypes(db, {schemaName: 'pg_catalog'});
  const groupedTypes = builtInTypes.reduce<{[key: string]: Type[]}>(
    (result, ty) => {
      const category = Object.keys(TypeCateogry).find(
        c => (TypeCateogry as any)[c] === ty.category,
      )!;
      result[category] = (result[category] || []).concat([ty]);
      return result;
    },
    {},
  );
  expect(
    Object.keys(groupedTypes)
      .sort()
      .reduce<{[key: string]: string[]}>((result, key) => {
        return {
          ...result,
          [key]: groupedTypes[key].map(
            ty =>
              `${ty.typeID} = ${ty.typeName}` +
              ('subtypeName' in ty && ty.subtypeName
                ? `<${ty.subtypeName}>`
                : '') +
              (ty.comment ? ' // ' + ty.comment : ''),
          ),
        };
      }, {}),
  ).toMatchSnapshot();

  const PgDataTypeIDsEnum = [
    '// autogenerated by test suite of pg-schema',
    '',
    'enum PgDataTypeID {',
  ];
  Object.keys(groupedTypes).forEach((groupName, i) => {
    if (i !== 0) PgDataTypeIDsEnum.push('');
    PgDataTypeIDsEnum.push(`  // === ${groupName} ===`);
    groupedTypes[groupName].forEach(type => {
      PgDataTypeIDsEnum.push('');
      const commentLines = [];
      if (type.comment) {
        commentLines.push(type.comment);
      }
      if (type.kind === TypeKind.Array) {
        commentLines.push(`Array<${type.subtypeName}>`);
      }
      if (commentLines.length) {
        PgDataTypeIDsEnum.push(`  /**`);
        commentLines.forEach((commentLine, j) => {
          if (j !== 0) {
            PgDataTypeIDsEnum.push(`   *`);
          }
          PgDataTypeIDsEnum.push(`   * ${commentLine}`);
        });
        PgDataTypeIDsEnum.push(`   */`);
      }
      PgDataTypeIDsEnum.push(`  ${type.typeName} = ${type.typeID},`);
    });
  });
  PgDataTypeIDsEnum.push(`}`);
  PgDataTypeIDsEnum.push(``);
  PgDataTypeIDsEnum.push(`export default PgDataTypeID;`);
  PgDataTypeIDsEnum.push(`module.exports = PgDataTypeID;`);
  PgDataTypeIDsEnum.push(`module.exports.default = PgDataTypeID;`);
  PgDataTypeIDsEnum.push(``);
  writeIfDifferent(
    __dirname + '/../../../pg-data-type-id/src/index.ts',
    PgDataTypeIDsEnum.join('\n'),
  );

  const pgTypes = require('pg-types/lib/textParsers');
  const mapping = new Map<number, unknown>();
  const reverseMapping = new Map<unknown, number[]>();
  pgTypes.init((id: number, parser: unknown) => {
    mapping.set(id, parser);
    const m = reverseMapping.get(parser) || [];
    reverseMapping.set(parser, [...m, id]);
  });
  const typeMappingLines: string[] = [];
  mapping.forEach((parser, id) => {
    const allIDs = reverseMapping.get(parser) || [];
    const idsWithMapping = allIDs.filter(typeID => typeID in typeMappings);
    if (idsWithMapping.length === 0) {
      throw new Error(
        'There is no mapping for: ' +
          allIDs.map(typeID => DataTypeID[typeID]).join(', '),
      );
    }
    if (idsWithMapping.length > 1) {
      throw new Error(
        'There is ambiguity between: ' +
          idsWithMapping.map(typeID => DataTypeID[typeID]).join(', '),
      );
    }
    typeMappingLines.push(
      `  [DataTypeID.${DataTypeID[id]}]: '${
        (typeMappings as any)[idsWithMapping[0]]
      }',`,
    );
  });
  const DefaultTypeScriptMapping = [
    `// autognerated by getTypes test`,
    ``,
    `import {DataTypeID} from '@databases/pg';`,
    ``,
    `const DefaultTypeScriptMapping: {[key in DataTypeID]?: string} = {`,
    ...typeMappingLines.sort(),
    `};`,
    ``,
    `export default DefaultTypeScriptMapping;`,
  ];
  writeIfDifferent(
    __dirname + '/../DefaultTypeScriptMapping.ts',
    DefaultTypeScriptMapping.join('\n'),
  );
});

test('get custom types', async () => {
  await db.query(sql`CREATE SCHEMA gettypes`);
  await db.query(
    sql`
      CREATE TYPE gettypes.currency AS ENUM('USD', 'GBP');
      COMMENT ON TYPE gettypes.currency IS 'Three character currency code';

      CREATE DOMAIN gettypes.email AS TEXT CHECK (VALUE ~ '^.+@.+$');
      COMMENT ON TYPE gettypes.email IS 'An email address';

      CREATE TYPE gettypes.money_with_currency AS (
        value MONEY,
        currency gettypes.currency
      );
      COMMENT ON TYPE gettypes.money_with_currency IS 'A monetary value with currency';
    `,
  );
  await db.query(
    sql`
      CREATE TABLE gettypes.tab (
        email gettypes.email NOT NULL PRIMARY KEY,
        money gettypes.money_with_currency
      );
    `,
  );
  await db.query(
    sql`
      INSERT INTO gettypes.tab (email, money) VALUES (${'forbes@lindesay.co.uk'}, ROW (${10}, 'USD'))
    `,
  );
  expect(await db.query(sql`SELECT * FROM gettypes.tab`))
    .toMatchInlineSnapshot(`
Array [
  Object {
    "email": "forbes@lindesay.co.uk",
    "money": "($10.00,USD)",
  },
]
`);
  expect(
    (await getTypes(db, {schemaName: 'gettypes'})).map(t => {
      const result = {
        ...t,
        schemaID: typeof t.schemaID === 'number' ? '<oid>' : t.schemaID,
        typeID: typeof t.typeID === 'number' ? '<oid>' : t.typeID,
      };
      if ('subtypeID' in result && typeof result.subtypeID === 'number') {
        result.subtypeID = '<oid>' as any;
      }
      if ('basetypeID' in result && typeof result.basetypeID === 'number') {
        result.basetypeID = '<oid>' as any;
      }
      if ('classID' in result && typeof result.classID === 'number') {
        result.classID = '<oid>' as any;
      }
      if ('attributes' in result) {
        result.attributes = result.attributes.map(a => ({
          ...a,
          classID: typeof a.classID === 'number' ? '<oid>' : a.classID,
          schemaID: typeof a.schemaID === 'number' ? '<oid>' : a.schemaID,
          typeID: typeof a.typeID === 'number' ? '<oid>' : a.typeID,
        })) as any[];
      }
      return result;
    }),
  ).toMatchInlineSnapshot(`
Array [
  Object {
    "category": "E",
    "comment": "Three character currency code",
    "kind": "e",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "typeID": "<oid>",
    "typeName": "currency",
    "values": Array [
      "USD",
      "GBP",
    ],
  },
  Object {
    "category": "A",
    "comment": null,
    "kind": "array",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "subtypeID": "<oid>",
    "subtypeName": "currency",
    "typeID": "<oid>",
    "typeName": "_currency",
  },
  Object {
    "basetypeID": "<oid>",
    "basetypeName": "text",
    "category": "S",
    "comment": "An email address",
    "kind": "d",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "typeID": "<oid>",
    "typeName": "email",
  },
  Object {
    "attributes": Array [
      Object {
        "attributeName": "value",
        "attributeNumber": 1,
        "classID": "<oid>",
        "className": "money_with_currency",
        "comment": null,
        "default": null,
        "hasDefault": false,
        "notNull": false,
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeLength": -1,
      },
      Object {
        "attributeName": "currency",
        "attributeNumber": 2,
        "classID": "<oid>",
        "className": "money_with_currency",
        "comment": null,
        "default": null,
        "hasDefault": false,
        "notNull": false,
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeLength": -1,
      },
    ],
    "category": "C",
    "classID": "<oid>",
    "comment": "A monetary value with currency",
    "kind": "c",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "typeID": "<oid>",
    "typeName": "money_with_currency",
  },
  Object {
    "category": "A",
    "comment": null,
    "kind": "array",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "subtypeID": "<oid>",
    "subtypeName": "money_with_currency",
    "typeID": "<oid>",
    "typeName": "_money_with_currency",
  },
  Object {
    "attributes": Array [
      Object {
        "attributeName": "email",
        "attributeNumber": 1,
        "classID": "<oid>",
        "className": "tab",
        "comment": null,
        "default": null,
        "hasDefault": false,
        "notNull": true,
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeLength": -1,
      },
      Object {
        "attributeName": "money",
        "attributeNumber": 2,
        "classID": "<oid>",
        "className": "tab",
        "comment": null,
        "default": null,
        "hasDefault": false,
        "notNull": false,
        "schemaID": "<oid>",
        "schemaName": "gettypes",
        "typeID": "<oid>",
        "typeLength": -1,
      },
    ],
    "category": "C",
    "classID": "<oid>",
    "comment": null,
    "kind": "c",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "typeID": "<oid>",
    "typeName": "tab",
  },
  Object {
    "category": "A",
    "comment": null,
    "kind": "array",
    "schemaID": "<oid>",
    "schemaName": "gettypes",
    "subtypeID": "<oid>",
    "subtypeName": "tab",
    "typeID": "<oid>",
    "typeName": "_tab",
  },
]
`);
});
