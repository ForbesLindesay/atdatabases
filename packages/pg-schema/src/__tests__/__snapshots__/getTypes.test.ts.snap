// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get built in types 1`] = `
Object {
  "Array": Array [
    "22 = int2vector<int2> // array of int2, used in system tables",
    "30 = oidvector<oid> // array of oids, used in system tables",
    "143 = _xml<xml>",
    "199 = _json<json>",
    "629 = _line<line>",
    "719 = _circle<circle>",
    "791 = _money<money>",
    "1000 = _bool<bool>",
    "1001 = _bytea<bytea>",
    "1002 = _char<char>",
    "1003 = _name<name>",
    "1005 = _int2<int2>",
    "1006 = _int2vector<int2vector>",
    "1007 = _int4<int4>",
    "1008 = _regproc<regproc>",
    "1009 = _text<text>",
    "1028 = _oid<oid>",
    "1010 = _tid<tid>",
    "1011 = _xid<xid>",
    "1012 = _cid<cid>",
    "1013 = _oidvector<oidvector>",
    "1014 = _bpchar<bpchar>",
    "1015 = _varchar<varchar>",
    "1016 = _int8<int8>",
    "1017 = _point<point>",
    "1018 = _lseg<lseg>",
    "1019 = _path<path>",
    "1020 = _box<box>",
    "1021 = _float4<float4>",
    "1022 = _float8<float8>",
    "1023 = _abstime<abstime>",
    "1024 = _reltime<reltime>",
    "1025 = _tinterval<tinterval>",
    "1027 = _polygon<polygon>",
    "1034 = _aclitem<aclitem>",
    "1040 = _macaddr<macaddr>",
    "775 = _macaddr8<macaddr8>",
    "1041 = _inet<inet>",
    "651 = _cidr<cidr>",
    "1263 = _cstring<cstring>",
    "1115 = _timestamp<timestamp>",
    "1182 = _date<date>",
    "1183 = _time<time>",
    "1185 = _timestamptz<timestamptz>",
    "1187 = _interval<interval>",
    "1231 = _numeric<numeric>",
    "1270 = _timetz<timetz>",
    "1561 = _bit<bit>",
    "1563 = _varbit<varbit>",
    "2201 = _refcursor<refcursor>",
    "2207 = _regprocedure<regprocedure>",
    "2208 = _regoper<regoper>",
    "2209 = _regoperator<regoperator>",
    "2210 = _regclass<regclass>",
    "2211 = _regtype<regtype>",
    "4097 = _regrole<regrole>",
    "4090 = _regnamespace<regnamespace>",
    "2951 = _uuid<uuid>",
    "3221 = _pg_lsn<pg_lsn>",
    "3643 = _tsvector<tsvector>",
    "3644 = _gtsvector<gtsvector>",
    "3645 = _tsquery<tsquery>",
    "3735 = _regconfig<regconfig>",
    "3770 = _regdictionary<regdictionary>",
    "3807 = _jsonb<jsonb>",
    "2949 = _txid_snapshot<txid_snapshot>",
    "3905 = _int4range<int4range>",
    "3907 = _numrange<numrange>",
    "3909 = _tsrange<tsrange>",
    "3911 = _tstzrange<tstzrange>",
    "3913 = _daterange<daterange>",
    "3927 = _int8range<int8range>",
  ],
  "BitString": Array [
    "1560 = bit // fixed-length bit string",
    "1562 = varbit // variable-length bit string",
  ],
  "Boolean": Array [
    "16 = bool // boolean, 'true'/'false'",
  ],
  "Composite": Array [
    "71 = pg_type",
    "75 = pg_attribute",
    "81 = pg_proc",
    "83 = pg_class",
    "10000 = pg_attrdef",
    "10001 = pg_constraint",
    "10002 = pg_inherits",
    "10003 = pg_index",
    "10004 = pg_operator",
    "10005 = pg_opfamily",
    "10006 = pg_opclass",
    "10130 = pg_am",
    "10131 = pg_amop",
    "10841 = pg_amproc",
    "11253 = pg_language",
    "11254 = pg_largeobject_metadata",
    "11255 = pg_largeobject",
    "11256 = pg_aggregate",
    "11257 = pg_statistic_ext",
    "11258 = pg_statistic",
    "11259 = pg_rewrite",
    "11260 = pg_trigger",
    "11261 = pg_event_trigger",
    "11262 = pg_description",
    "11263 = pg_cast",
    "11483 = pg_enum",
    "11484 = pg_namespace",
    "11485 = pg_conversion",
    "11486 = pg_depend",
    "1248 = pg_database",
    "11487 = pg_db_role_setting",
    "11488 = pg_tablespace",
    "11489 = pg_pltemplate",
    "2842 = pg_authid",
    "2843 = pg_auth_members",
    "11490 = pg_shdepend",
    "11491 = pg_shdescription",
    "11492 = pg_ts_config",
    "11493 = pg_ts_config_map",
    "11494 = pg_ts_dict",
    "11495 = pg_ts_parser",
    "11496 = pg_ts_template",
    "11497 = pg_extension",
    "11498 = pg_foreign_data_wrapper",
    "11499 = pg_foreign_server",
    "11500 = pg_user_mapping",
    "11501 = pg_foreign_table",
    "11502 = pg_policy",
    "11503 = pg_replication_origin",
    "11504 = pg_default_acl",
    "11505 = pg_init_privs",
    "11506 = pg_seclabel",
    "4066 = pg_shseclabel",
    "11507 = pg_collation",
    "11508 = pg_partitioned_table",
    "11509 = pg_range",
    "11510 = pg_transform",
    "11511 = pg_sequence",
    "11512 = pg_publication",
    "11513 = pg_publication_rel",
    "6101 = pg_subscription",
    "11514 = pg_subscription_rel",
    "11528 = pg_roles",
    "11532 = pg_shadow",
    "11536 = pg_group",
    "11539 = pg_user",
    "11542 = pg_policies",
    "11546 = pg_rules",
    "11550 = pg_views",
    "11554 = pg_tables",
    "11558 = pg_matviews",
    "11562 = pg_indexes",
    "11566 = pg_sequences",
    "11570 = pg_stats",
    "11574 = pg_publication_tables",
    "11578 = pg_locks",
    "11581 = pg_cursors",
    "11584 = pg_available_extensions",
    "11587 = pg_available_extension_versions",
    "11590 = pg_prepared_xacts",
    "11594 = pg_prepared_statements",
    "11597 = pg_seclabels",
    "11601 = pg_settings",
    "11606 = pg_file_settings",
    "11609 = pg_hba_file_rules",
    "11612 = pg_timezone_abbrevs",
    "11615 = pg_timezone_names",
    "11618 = pg_config",
    "11621 = pg_stat_all_tables",
    "11625 = pg_stat_xact_all_tables",
    "11629 = pg_stat_sys_tables",
    "11633 = pg_stat_xact_sys_tables",
    "11636 = pg_stat_user_tables",
    "11640 = pg_stat_xact_user_tables",
    "11643 = pg_statio_all_tables",
    "11647 = pg_statio_sys_tables",
    "11650 = pg_statio_user_tables",
    "11653 = pg_stat_all_indexes",
    "11657 = pg_stat_sys_indexes",
    "11660 = pg_stat_user_indexes",
    "11663 = pg_statio_all_indexes",
    "11667 = pg_statio_sys_indexes",
    "11670 = pg_statio_user_indexes",
    "11673 = pg_statio_all_sequences",
    "11677 = pg_statio_sys_sequences",
    "11680 = pg_statio_user_sequences",
    "11683 = pg_stat_activity",
    "11687 = pg_stat_replication",
    "11691 = pg_stat_wal_receiver",
    "11694 = pg_stat_subscription",
    "11697 = pg_stat_ssl",
    "11700 = pg_replication_slots",
    "11704 = pg_stat_database",
    "11707 = pg_stat_database_conflicts",
    "11710 = pg_stat_user_functions",
    "11714 = pg_stat_xact_user_functions",
    "11718 = pg_stat_archiver",
    "11721 = pg_stat_bgwriter",
    "11724 = pg_stat_progress_vacuum",
    "11728 = pg_user_mappings",
    "11732 = pg_replication_origin_status",
  ],
  "DateTime": Array [
    "702 = abstime // absolute, limited-range date and time (Unix system time)",
    "1082 = date // date",
    "1083 = time // time of day",
    "1114 = timestamp // date and time",
    "1184 = timestamptz // date and time with time zone",
    "1266 = timetz // time of day with time zone",
  ],
  "Geometric": Array [
    "600 = point<float8> // geometric point '(x, y)'",
    "601 = lseg<point> // geometric line segment '(pt1,pt2)'",
    "602 = path // geometric path '(pt1,...)'",
    "603 = box<point> // geometric box '(lower left,upper right)'",
    "604 = polygon // geometric polygon '(pt1,...)'",
    "628 = line<float8> // geometric line",
    "718 = circle // geometric circle '(center,radius)'",
  ],
  "NetworkAddress": Array [
    "869 = inet // IP address/netmask, host address, netmask optional",
    "650 = cidr // network IP address/netmask, network address",
  ],
  "Numeric": Array [
    "20 = int8 // ~18 digit integer, 8-byte storage",
    "21 = int2 // -32 thousand to 32 thousand, 2-byte storage",
    "23 = int4 // -2 billion to 2 billion integer, 4-byte storage",
    "24 = regproc // registered procedure",
    "26 = oid // object identifier(oid), maximum 4 billion",
    "700 = float4 // single-precision floating point number, 4-byte storage",
    "701 = float8 // double-precision floating point number, 8-byte storage",
    "790 = money // monetary amounts, $d,ddd.cc",
    "1700 = numeric // numeric(precision, decimal), arbitrary precision number",
    "2202 = regprocedure // registered procedure (with args)",
    "2203 = regoper // registered operator",
    "2204 = regoperator // registered operator (with args)",
    "2205 = regclass // registered class",
    "2206 = regtype // registered type",
    "4096 = regrole // registered role",
    "4089 = regnamespace // registered namespace",
    "3734 = regconfig // registered text search configuration",
    "3769 = regdictionary // registered text search dictionary",
  ],
  "PseudoTypes": Array [
    "32 = pg_ddl_command // internal type for passing CollectedCommand",
    "2249 = record",
    "2287 = _record",
    "2275 = cstring",
    "2276 = any",
    "2277 = anyarray",
    "2278 = void",
    "2279 = trigger",
    "3838 = event_trigger",
    "2280 = language_handler",
    "2281 = internal",
    "2282 = opaque",
    "2283 = anyelement",
    "2776 = anynonarray",
    "3500 = anyenum",
    "3115 = fdw_handler",
    "325 = index_am_handler",
    "3310 = tsm_handler",
    "3831 = anyrange",
  ],
  "Range": Array [
    "3904 = int4range // range of integers",
    "3906 = numrange // range of numerics",
    "3908 = tsrange // range of timestamps without time zone",
    "3910 = tstzrange // range of timestamps with time zone",
    "3912 = daterange // range of dates",
    "3926 = int8range // range of bigints",
  ],
  "String": Array [
    "18 = char // single character",
    "19 = name<char> // 63-byte type for storing system identifiers",
    "25 = text // variable-length string, no limit specified",
    "194 = pg_node_tree // string representing an internal node tree",
    "3361 = pg_ndistinct // multivariate ndistinct coefficients",
    "3402 = pg_dependencies // multivariate dependencies",
    "1042 = bpchar // char(length), blank-padded string, fixed storage length",
    "1043 = varchar // varchar(length), non-blank-padded string, variable storage length",
  ],
  "Timespan": Array [
    "703 = reltime // relative, limited-range time interval (Unix delta time)",
    "704 = tinterval // (abstime,abstime), time interval",
    "1186 = interval // @ <number> <units>, time interval",
  ],
  "Unknown": Array [
    "705 = unknown",
  ],
  "UserDefined": Array [
    "17 = bytea // variable-length string, binary values escaped",
    "27 = tid // (block, offset), physical location of tuple",
    "28 = xid // transaction id",
    "29 = cid // command identifier type, sequence in transaction id",
    "114 = json",
    "142 = xml // XML content",
    "210 = smgr // storage manager",
    "829 = macaddr // XX:XX:XX:XX:XX:XX, MAC address",
    "774 = macaddr8 // XX:XX:XX:XX:XX:XX:XX:XX, MAC address",
    "1033 = aclitem // access control list",
    "1790 = refcursor // reference to cursor (portal name)",
    "2950 = uuid // UUID datatype",
    "3220 = pg_lsn // PostgreSQL LSN datatype",
    "3614 = tsvector // text representation for text search",
    "3642 = gtsvector // GiST index internal text representation for text search",
    "3615 = tsquery // query representation for text search",
    "3802 = jsonb // Binary JSON",
    "2970 = txid_snapshot // txid snapshot",
  ],
}
`;
